### PR TITLE
[lit-html] Add cache directive

### PR DESCRIPTION
**Pull request by justinfagnani**
Also adds NodePart.detach() and NodePart.restore()

### CODE REVIEW

File: packages/lit-html/src/lib/lit-html.ts [hunk 1]
```diff
@@ -569,6 +582,28 @@ export class NodePart {
     public options: RenderOptions | undefined
   ) {}
 
+  detach(): NodePartState {
```
Comment (sorvell) [inline packages/lit-html/src/lib/lit-html.ts:585]: Moving this out of the core would be ideal since it's only needed for some directives. 

Maybe the directive base class could have these methods and maybe that could be a separate import.

File: packages/lit-html/src/lib/directives/cache.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ * `
+ * ```
+ */
+export const cache = directive(
+  class extends Directive {
+    templateCache = new WeakMap<TemplateStringsArray, NodePartState>();
+    value?: TemplateResult;
+
+    render(v: unknown) {
+      if ((v as TemplateResult)._$litType$ !== undefined) {
```
Comment (jridgewell) [inline packages/lit-html/src/lib/directives/cache.ts:43]: `v` can be nullish here, right?
```suggestion
      if ((v as TemplateResult | null)?._$litType$ !== undefined) {
```

File: packages/lit-html/src/lib/directives/cache.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      }
+      return v;
+    }
+
+    update(part: NodePart, [v]: Parameters<this['render']>) {
+      // If the new value is not a TemplateResult from the same Template as the
+      // previous value, move the nodes from the DOM into the cache.
+      if (
+        this.value !== undefined &&
+        this.value.strings !== (v as TemplateResult).strings
```
Comment (jridgewell) [inline packages/lit-html/src/lib/directives/cache.ts:56]: ditto.

File: packages/lit-html/src/lib/lit-html.ts (hunk 1 above)
Comment (justinfagnani) [reply packages/lit-html/src/lib/lit-html.ts:585]: Moving these to a separate module

File: packages/lit-html/src/lib/directives/cache.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+ * ```
+ */
+export const cache = directive(
+  class extends Directive {
+    templateCache = new WeakMap<TemplateStringsArray, NodePartState>();
+    value?: TemplateResult;
+
+    render(v: unknown) {
+      if ((v as TemplateResult)._$litType$ !== undefined) {
+        this.value = v as TemplateResult;
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/lit-html/src/lib/directives/cache.ts:44]: render() is here for SSR. We'll discuss the directive lifecycle more...

**Comment by justinfagnani** [comment]
I'm going to close this and #1220 in favor of a PR that puts part mutation in a helper module.
